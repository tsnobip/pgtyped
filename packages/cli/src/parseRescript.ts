import { parseSQLFile } from '@pgtyped/parser';
import { SQLParseResult } from '@pgtyped/parser/lib/loader/sql';

export function parseCode(fileContent: string): SQLParseResult {
  if (!fileContent.includes('%sql')) {
    return {
      queries: [],
      events: [],
    };
  }

  // Replace with more robust @rescript/tools CLI usage when that package ships linuxarm64 binary.
  const regex = /%sql(?:\.\w+)?\(`([^`]*)`\)/g;
  let match;
  const queries = [];

  while ((match = regex.exec(fileContent)) !== null) {
    let query = match[1].trim();
    if (!query.endsWith(';')) {
      query += ';';
    }

    if (!query.includes('@name')) {
      // Handle potentially existing doc comment
      if (query.trim().startsWith('/*')) {
        const lines = query.split('\n');

        let comment = `/* @name Query${queries.length + 1}\n`;
        for (let i = 0; i <= lines.length - 1; i += 1) {
          const line = lines[i].trim().replace('/*', '');
          comment += line + '\n';
          if (line.endsWith('*/')) {
            break;
          }
        }
        query = `${comment}\n${query}`;
      } else {
        query = `/* @name Query${queries.length + 1} */\n${query}`;
      }
    }

    queries.push(query);
  }

  const asSql = queries.join('\n\n');

  const res = parseSQLFile(asSql);
  return res;
}
