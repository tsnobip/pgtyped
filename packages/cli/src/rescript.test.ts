import { parseCode } from './parseRescript.js';

test('parser finds string template in correct file', () => {
  const fileContent = `
    let query = %sql.one(\`
    /* @name FindBookById */
    SELECT * FROM books WHERE id = :id; 
    \`);
  `;

  const result = parseCode(fileContent);
  expect(result).toMatchSnapshot();
});

test('inserts trailing semicolon when needed', () => {
  const fileContent = `
    let query = %sql.one(\`
    /* @name FindBookById */
    SELECT * FROM books WHERE id = :id 
    \`);

    let queryMany = %sql.many(\`
    /* @name FindBooks */
    SELECT * FROM books
    \`);
  `;

  const result = parseCode(fileContent);
  expect(result).toMatchSnapshot();
});

test('inserts @name when needed', () => {
  const fileContent = `
    let query = %sql.one(\`
    /* @name FindBookById */
    SELECT * FROM books WHERE id = :id 
    \`);

    let queryMany = %sql.many(\`
    SELECT * FROM books
    \`);

    let queryExpect = %sql.expectOne(\`
    SELECT * FROM books WHERE id = :id
    \`);
  `;

  const result = parseCode(fileContent);
  expect(result).toMatchSnapshot();
});

test('preserves @param when inserting @name', () => {
  const fileContent = `
    let query = %sql.one(\`
    /*
      @param notification -> (payload, user_id, type)
    */
    INSERT INTO notifications (payload, user_id, type) VALUES :notification
    \`);
  `;

  const result = parseCode(fileContent);
  expect(result).toMatchSnapshot();
});
