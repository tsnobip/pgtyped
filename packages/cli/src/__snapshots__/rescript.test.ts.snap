// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`inserts @name when needed 1`] = `
Object {
  "events": Array [],
  "queries": Array [
    Object {
      "name": "FindBookById",
      "params": Array [
        Object {
          "codeRefs": Object {
            "used": Array [
              Object {
                "a": 61,
                "b": 62,
                "col": 36,
                "line": 2,
              },
            ],
          },
          "name": "id",
          "required": false,
          "transform": Object {
            "type": "scalar",
          },
        },
      ],
      "statement": Object {
        "body": "SELECT * FROM books WHERE id = :id",
        "loc": Object {
          "a": 29,
          "b": 62,
          "col": 4,
          "line": 2,
        },
      },
      "usedParamSet": Object {
        "id": true,
      },
    },
    Object {
      "name": "Query2",
      "params": Array [],
      "statement": Object {
        "body": "SELECT * FROM books",
        "loc": Object {
          "a": 85,
          "b": 103,
          "col": 0,
          "line": 5,
        },
      },
      "usedParamSet": Object {},
    },
    Object {
      "name": "Query3",
      "params": Array [
        Object {
          "codeRefs": Object {
            "used": Array [
              Object {
                "a": 158,
                "b": 159,
                "col": 32,
                "line": 8,
              },
            ],
          },
          "name": "id",
          "required": false,
          "transform": Object {
            "type": "scalar",
          },
        },
      ],
      "statement": Object {
        "body": "SELECT * FROM books WHERE id = :id",
        "loc": Object {
          "a": 126,
          "b": 159,
          "col": 0,
          "line": 8,
        },
      },
      "usedParamSet": Object {
        "id": true,
      },
    },
  ],
}
`;

exports[`inserts trailing semicolon when needed 1`] = `
Object {
  "events": Array [],
  "queries": Array [
    Object {
      "name": "FindBookById",
      "params": Array [
        Object {
          "codeRefs": Object {
            "used": Array [
              Object {
                "a": 61,
                "b": 62,
                "col": 36,
                "line": 2,
              },
            ],
          },
          "name": "id",
          "required": false,
          "transform": Object {
            "type": "scalar",
          },
        },
      ],
      "statement": Object {
        "body": "SELECT * FROM books WHERE id = :id",
        "loc": Object {
          "a": 29,
          "b": 62,
          "col": 4,
          "line": 2,
        },
      },
      "usedParamSet": Object {
        "id": true,
      },
    },
    Object {
      "name": "FindBooks",
      "params": Array [],
      "statement": Object {
        "body": "SELECT * FROM books",
        "loc": Object {
          "a": 92,
          "b": 110,
          "col": 4,
          "line": 5,
        },
      },
      "usedParamSet": Object {},
    },
  ],
}
`;

exports[`parser finds string template in correct file 1`] = `
Object {
  "events": Array [],
  "queries": Array [
    Object {
      "name": "FindBookById",
      "params": Array [
        Object {
          "codeRefs": Object {
            "used": Array [
              Object {
                "a": 61,
                "b": 62,
                "col": 36,
                "line": 2,
              },
            ],
          },
          "name": "id",
          "required": false,
          "transform": Object {
            "type": "scalar",
          },
        },
      ],
      "statement": Object {
        "body": "SELECT * FROM books WHERE id = :id",
        "loc": Object {
          "a": 29,
          "b": 62,
          "col": 4,
          "line": 2,
        },
      },
      "usedParamSet": Object {
        "id": true,
      },
    },
  ],
}
`;

exports[`preserves @param when inserting @name 1`] = `
Object {
  "events": Array [],
  "queries": Array [
    Object {
      "name": "Query1",
      "params": Array [
        Object {
          "codeRefs": Object {
            "defined": Object {
              "a": 24,
              "b": 35,
              "col": 7,
              "line": 3,
            },
            "used": Array [
              Object {
                "a": 196,
                "b": 207,
                "col": 63,
                "line": 9,
              },
            ],
          },
          "name": "notification",
          "required": false,
          "transform": Object {
            "keys": Array [
              Object {
                "name": "payload",
                "required": false,
              },
              Object {
                "name": "user_id",
                "required": false,
              },
              Object {
                "name": "type",
                "required": false,
              },
            ],
            "type": "pick_tuple",
          },
        },
      ],
      "statement": Object {
        "body": "                                                               
    INSERT INTO notifications (payload, user_id, type) VALUES :notification",
        "loc": Object {
          "a": 69,
          "b": 207,
          "col": 0,
          "line": 6,
        },
      },
      "usedParamSet": Object {
        "notification": true,
      },
    },
  ],
}
`;
